src/sentry/search/events/builder.py

```diff
@@ -1744,32 +1729,65 @@ def run_query(self, referrer: str, use_cache: bool = False) -> Any:
 
 class TimeseriesMetricQueryBuilder(MetricsQueryBuilder):
     time_alias = "time"
-    time_column = AliasedExpression(Column("timestamp"), time_alias)
 
     def __init__(
         self,
         params: ParamsType,
-        granularity: int,
+        interval: int,
         query: Optional[str] = None,
         selected_columns: Optional[List[str]] = None,
         functions_acl: Optional[List[str]] = None,
     ):
         super().__init__(
             params=params,
-            granularity=granularity,
             query=query,
             selected_columns=selected_columns,
             auto_fields=False,
             functions_acl=functions_acl,
         )
 
+        self.time_column = self.resolve_time_column(interval)
+
         # This is a timeseries, the groupby will always be time
         self.groupby = [self.time_column]
 
-    def resolve_granularity(self, granularity: Optional[int] = None) -> Optional[Granularity]:
-        if granularity is None:
-            raise InvalidSearchQuery("Cannot query a timeseries without granularity")
-        return Granularity(granularity)
+    def resolve_time_column(self, interval: int) -> Function:
+        """Need to round the timestamp to the interval requested
+
+        We commonly use interval & granularity interchangeably, but in the case of the metrics dataset they must be
+        considered as two separate things. The reason being the way we store metrics will rarely align with the
+        start&end of the query.
+        This means that we'll need to select granularity for data accuracy, and then use the clickhouse
+        toStartOfInterval function to group results by their displayed interval
+
+        eg.
+        See test_builder.test_run_query_with_hour_interval for this in test form
+        we have a query from yesterday at 15:30 -> today at 15:30
+        there is 1 event at 15:45
+        and we want the timeseries displayed at 1 hour intervals
+
+        The event is in the quantized hour-aligned metrics bucket of 15:00, since the bounds of the query are
+        (Yesterday 15:30, Today 15:30) the condition > Yesterday 15:30 means using the hour-aligned bucket you'd
+        miss that event.
+
+        So instead in this case we want the minute-aligned bucket, while rounding timestamp to the hour, so we'll
+        only get data that is relevant because of the timestamp filters. And Snuba will merge the datasketches for
+        us to get correct data.
+        """
+        if interval < 10:
+            raise IncompatibleMetricsQuery(
+                "Interval must be at least 10s because our smallest granularity is 10s"
+            )
+
+        return Function(
```

Comment:
Okay, so I think this makes sense to me. We're doing two things:
1. Using a more precise granularity (60 seconds instead on 1 hour for example) in cases where the start/end time doesn't align with the bucket intervals. We're doing this so we don't miss out on data that would be otherwise discarded by the hourly buckets due to start time being in the middle of the bucket.
2. We're grouping the results by timestamps rounded to the start of the interval, based on the interval from the incoming request, so that we get the response in the expected interval instead of the more granular results we would have otherwise gotten as we tried to increase data accuracy.